Working on fixes for Android and Vulkan

BastiaanOlij opened the pull request:
Fixes a few crash related issues on Android.

- Correct vulkan context is now created again instead of an OpenGL context (related setting was checked incorrectly)
- Added checks for valid ENV in wrapper
- Added a solution to make sure our thread is properly bound to the JavaVM for calling

That last one needs more work, I did the bare minimum to make it work without a big rewrite that would have much more far reaching consequences. Ideally we should have code that binds on first use (which we do now) but cleans up when the thread terminates. The current solution is workable and basically does what its always done but it can be made better.

BastiaanOlij committed:
Fixes on android:
- creating Vulkan context instead of OpenGL
- checking for validity of ENV in wrapper classes
- fix for access to JavaVM from threads

BastiaanOlij commented:
Now ends up crashing here:
```
03-09 21:26:46.945 16907 16977 I godot   : Godot Engine v4.0.dev.custom_build.e1804bfb7 - https://godotengine.org
03-09 21:26:46.948  3817  3817 D QtiCarrierConfigHelper: WARNING, no carrier configs on phone Id: 0
03-09 21:26:46.969  4279 16252 I Icing   : Usage reports ok 0, Failed Usage reports 0, indexed 0, rejected 0
03-09 21:26:46.971 16907 16977 D vulkan  : searching for layers in '/data/app/org.godotengine.testandroid4-hxlMCycOzO7d5qoKpfsOBA==/lib/arm64'
03-09 21:26:46.972 16907 16977 D vulkan  : searching for layers in '/data/app/org.godotengine.testandroid4-hxlMCycOzO7d5qoKpfsOBA==/base.apk!/lib/arm64-v8a'
03-09 21:26:46.975  4279 16252 I Icing   : Indexing com.google.android.gms-internal.3p:MobileApplication from com.google.android.gms using seqno 0
03-09 21:26:46.980 16907 16977 I AdrenoVK-0: QUALCOMM build          : 39a1dfd, Ic628754133
03-09 21:26:46.980 16907 16977 I AdrenoVK-0: Build Date              : 06/29/20
03-09 21:26:46.980 16907 16977 I AdrenoVK-0: Shader Compiler Version : EV031.29.00.13
03-09 21:26:46.980 16907 16977 I AdrenoVK-0: Local Branch            :
03-09 21:26:46.980 16907 16977 I AdrenoVK-0: Remote Branch           :
03-09 21:26:46.980 16907 16977 I AdrenoVK-0: Remote Branch           :
03-09 21:26:46.980 16907 16977 I AdrenoVK-0: Reconstruct Branch      :
03-09 21:26:46.980 16907 16977 I AdrenoVK-0: Build Config            : S P 8.0.16 AArch64
03-09 21:26:46.980 16907 16977 I AdrenoVK-0: Driver Path             : /vendor/lib64/hw/vulkan.bengal.so
03-09 21:26:46.981  4279 16252 I Icing   : Indexing com.google.android.gms-internal.3p:MobileApplication from com.google.android.gms using seqno 0
03-09 21:26:46.983 16907 16977 I godot   : Using Vulkan Device #0: Qualcomm - Adreno (TM) 610
03-09 21:26:46.994  4279 16252 I Icing   : Indexing com.google.android.gms-apps from com.google.android.gms
03-09 21:26:46.999  1853  7095 I LaunchCheckinHandler: Displayed org.godotengine.testandroid4/com.godot.game.GodotApp,cp,ca,966
03-09 21:26:46.999  1853  7095 I WindowManager: Window drawn AppWindowToken{174756 token=Token{548bd71 ActivityRecord{cba1f18 u0 org.godotengine.testandroid4/com.godot.game.GodotApp t29}}}
03-09 21:26:47.000  1853  7095 I WindowManager:   SURFACE show Surface(name=org.godotengine.testandroid4/com.godot.game.GodotApp)/@0xc968516 on display:0: org.godotengine.testandroid4/com.godot.game.GodotApp
--------- beginning of crash
03-09 21:26:47.001 16907 16977 F libc    : Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x0 in tid 16977 (VkThread), pid 16907 (ne.testandroid4)
03-09 21:26:47.003  1853  1954 I ActivityTaskManager: Displayed org.godotengine.testandroid4/com.godot.game.GodotApp: +1s52ms
03-09 21:26:47.009  1853  1853 V SettingsProvider: Notifying for 0: content://settings/secure/reminder_exp_learning_time_elapsed
03-09 21:26:47.034  4772  4772 I GoogleInputMethodService: GoogleInputMethodService.onFinishInput():3311
03-09 21:26:47.036  4772  4772 I GoogleInputMethodService: GoogleInputMethodService.onStartInput():1914
03-09 21:26:47.081  4279 16252 I Icing   : Indexing done com.google.android.gms-apps
03-09 21:26:47.082  4279 16252 I Icing   : Indexing com.google.android.gms-apps from com.google.android.gms
03-09 21:26:47.085  4279 16252 I Icing   : Indexing done com.google.android.gms-apps
03-09 21:26:47.090 16985 16985 I crash_dump64: obtaining output fd from tombstoned, type: kDebuggerdTombstone
03-09 21:26:47.091  1382  1382 I /system/bin/tombstoned: received crash request for pid 16977
03-09 21:26:47.093 16985 16985 I crash_dump64: performing dump of process 16907 (target tid = 16977)
03-09 21:26:47.103 16985 16985 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
03-09 21:26:47.103 16985 16985 F DEBUG   : Build fingerprint: 'motorola/guamp_retailn/guamp:10/QPX30.30-Q3-38-54/1fcf8:user/release-keys'
03-09 21:26:47.103 16985 16985 F DEBUG   : Revision: 'pvt'
03-09 21:26:47.103 16985 16985 F DEBUG   : ABI: 'arm64'
03-09 21:26:47.104 16985 16985 F DEBUG   : Timestamp: 2021-03-09 21:26:47+1100
03-09 21:26:47.104 16985 16985 F DEBUG   : pid: 16907, tid: 16977, name: VkThread  >>> org.godotengine.testandroid4 <<<
03-09 21:26:47.104 16985 16985 F DEBUG   : uid: 10240
03-09 21:26:47.104 16985 16985 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x0
03-09 21:26:47.104 16985 16985 F DEBUG   : Cause: null pointer dereference
03-09 21:26:47.104 16985 16985 F DEBUG   :     x0  0000000000000000  x1  0000000000002786  x2  0000007b7038aa70  x3  0000000000000000
03-09 21:26:47.104 16985 16985 F DEBUG   :     x4  0000007b0e5326e0  x5  0000007bfe3087f0  x6  0000000000000000  x7  0000000000000000
03-09 21:26:47.104 16985 16985 F DEBUG   :     x8  ffffff80ffffffd8  x9  0000007b09988d40  x10 0000007b09988d70  x11 83eb65540044ed54
03-09 21:26:47.104 16985 16985 F DEBUG   :     x12 0000007b09988de0  x13 0000000000156d58  x14 0000000000000030  x15 0040000000000000
03-09 21:26:47.104 16985 16985 F DEBUG   :     x16 0000007b0f0da328  x17 0000007bfa317280  x18 0000007b08ab6000  x19 0000007b0998a020
03-09 21:26:47.104 16985 16985 F DEBUG   :     x20 0000000000000001  x21 0000007b0f0dd000  x22 0000007b0998a020  x23 0000007b0f0e1000
03-09 21:26:47.104 16985 16985 F DEBUG   :     x24 0000007b0f0e1000  x25 0000007b0f0e1000  x26 0000007b6854bcb0  x27 0000000000000000
03-09 21:26:47.104 16985 16985 F DEBUG   :     x28 0000007b09989040  x29 0000007b09989040
03-09 21:26:47.104 16985 16985 F DEBUG   :     sp  0000007b09988cc0  lr  0000007b0c3c0610  pc  0000007b0c396a08
03-09 21:26:47.237 16985 16985 F DEBUG   :
03-09 21:26:47.237 16985 16985 F DEBUG   : backtrace:
03-09 21:26:47.237 16985 16985 F DEBUG   :       #00 pc 0000000000209a08  /data/app/org.godotengine.testandroid4-hxlMCycOzO7d5qoKpfsOBA==/lib/arm64/libgodot_android.so
03-09 21:26:47.237 16985 16985 F DEBUG   :       #01 pc 000000000023360c  /data/app/org.godotengine.testandroid4-hxlMCycOzO7d5qoKpfsOBA==/lib/arm64/libgodot_android.so
03-09 21:26:47.237 16985 16985 F DEBUG   :       #02 pc 000000000020cb64  /data/app/org.godotengine.testandroid4-hxlMCycOzO7d5qoKpfsOBA==/lib/arm64/libgodot_android.so (Java_org_godotengine_godot_GodotLib_step+84)
03-09 21:26:47.237 16985 16985 F DEBUG   :       #03 pc 0000000000140350  /apex/com.android.runtime/lib64/libart.so (art_quick_generic_jni_trampoline+144) (BuildId: dd6d53efa16333210280f1910dd28dbc)
03-09 21:26:47.237 16985 16985 F DEBUG   :       #04 pc 00000000001375b8  /apex/com.android.runtime/lib64/libart.so (art_quick_invoke_static_stub+568) (BuildId: dd6d53efa16333210280f1910dd28dbc)
03-09 21:26:47.237 16985 16985 F DEBUG   :       #05 pc 000000000014608c  /apex/com.android.runtime/lib64/libart.so (art::ArtMethod::Invoke(art::Thread*, unsigned int*, unsigned int, art::JValue*, char const*)+276) (BuildId: dd6d53efa16333210280f1910dd28dbc)
03-09 21:26:47.237 16985 16985 F DEBUG   :       #06 pc 00000000002e16a8  /apex/com.android.runtime/lib64/libart.so (art::interpreter::ArtInterpreterToCompiledCodeBridge(art::Thread*, art::ArtMethod*, art::ShadowFrame*, unsigned short, art::JValue*)+384) (BuildId: dd6d53efa16333210280f1910dd28dbc)
```

akien-mga commented:
You can build you Android template with `target=debug` to have a usable stacktrace.

BastiaanOlij commented:
Ok, compiling with `target=debug` and un-commenting the 3 `// doNotStrip '**/*.so'` lines in the gradle configuration gave me some proper debug info and I found it was `set_screen_orientation` that was crashing the system. Specifically it was `env` being `NULL`.

I've added a bunch of conditions here so we get proper logging. Further investigation shows this:
`03-10 11:50:11.758 13905 13905 I godot   : Java_org_godotengine_godot_GodotLib_initialize` 
That is a print I added at the start of `Java_org_godotengine_godot_GodotLib_initialize` and it shows that our pid is 13905. This is running on the main/starting thread of the application.

But when it goes wrong:
```
03-10 11:50:12.079 13905 13986 E godot   : ERROR: Condition "env == nullptr" is true.
03-10 11:50:12.079 13905 13986 E godot   :    at: set_screen_orientation (platform\android\java_godot_io_wrapper.cpp:176) - Condition "env == nullptr" is true.
```

Note how we're now running on pid 13986, Godot 4 makes much more use of threading and our render engine runs in separate threads. It looks like a thread is calling back into our GodotIO wrapper and it doesn't have access to a valid env.

I'm not sure how the threading model should work here, The JAVA VM is still a big black box for me.

@m4gr3d any ideas?

BastiaanOlij commented:
ok, so after reading up on things a little more, any thread that needs access to the Java VM should attach itself and will obtain an env pointer. Once the thread is done it should detach itself from the VM. 

Now in its original usage the `init_thread` and `term_thread` were part of a callback structure where the current thread is calling back into another thread, not even sure this is still used anymore. @reduz can you comment on this?

Anyway it's an ugly hack for now but I'm now checking if we have a valid environment on setup and get_jni_env and if not I'm attaching the current thread to the javaVM. It's allowing us to do the calls from other threads but its not the right solution, just a patch to see if we can get things up and running.

I think we need further discussion about the lifecycle of threads now that we are use more of them and making sure threads properly clean up any platform dependent things. As this has to be called from within the thread function we'll need to think of a nice callback for this. Maybe a cleanup_thread function pointer we register in the same way as we register the other pointers but they should be called at the end of the thread function. 

The lazy attaching of the VM makes sense to keep because only a handful of threads will ever communicate back into the Java side of life.

BastiaanOlij commented:
Ok compiler, you are drunk... when I set it to `%lu` you complain it has to be `%llu`, when I set it to `%llu` you complain it should be `%lu`....

Those prints are temporary anyway so ignore for now :)

BastiaanOlij commented:
I've made this ready for review because these 3 fixes are needed for continued testing of Android in Godot 4. Together with @m4gr3d fixes we're inching closer to a working Android build (I think us using SPIRV 1.1 features is the only real hurdle left).
But since I'll be busy for the next week or so building a toned down version of the Vulkan renderer I think further fixes should go into separate PRs and we should merge the fixes so far.

akien-mga commented:
Thanks!

akien-mga commented:
Cherry-picked for 3.3 via #47682 (aside from Vulkan changes), it was also relevant there after the Thread modernization changes.
